core/message.go

```diff
@@ -0,0 +1,116 @@
+package core
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/peak/s5cmd/log"
+	"github.com/peak/s5cmd/objurl"
+	"github.com/peak/s5cmd/storage"
+)
+
+// InfoMessage is a generic message structure for successful operations.
+type InfoMessage struct {
+	Operation   string            `json:"operation"`
+	Success     bool              `json:"success"`
+	Source      *objurl.ObjectURL `json:"source"`
+	Destination *objurl.ObjectURL `json:"destination,omitempty"`
+	Object      *storage.Object   `json:"object,omitempty"`
+}
+
+// String is the string representation of InfoMessage.
+func (u InfoMessage) String() string {
+	return fmt.Sprintf("%v %v", u.Operation, u.Source)
+}
+
+// JSON is the JSON representation of InfoMessage.
+func (u InfoMessage) JSON() string {
+	u.Success = true
+	bytes, _ := json.Marshal(u)
+	return string(bytes)
+}
+
+// ErrorMessage is a generic message structure for unsuccessful operations.
+type ErrorMessage struct {
+	Operation string `json:"operation,omitempty"`
+	Job       string `json:"job"`
+	Err       string `json:"error"`
+	Format    string `json:"-"`
+}
+
+// String is the string representation of ErrorMessage.
+func (e ErrorMessage) String() string {
+	err := cleanupSpaces(e.Err)
+	return fmt.Sprintf(e.Format, e.Job, err)
+}
+
+// JSON is the JSON representation of ErrorMessage.
+func (e ErrorMessage) JSON() string {
+	e.Err = cleanupSpaces(e.Err)
```

Comment:
i'm not sure if setting e.Err would be persistent across function calls. `e` is not a pointer receiver. Could you test?

Reply:
It is not needed since it is only needed in this function scope. But I found a better way. We can use `cleanupSpaces` method on `log` package.

Reply:
Sorry, I moved it into `newErrorMessage()`. We can clean up multi-line strings on the constructor.